libp2p/NodeTable.cpp

```diff
@@ -304,101 +308,117 @@ vector<shared_ptr<NodeEntry>> NodeTable::nearestNodeEntries(NodeID _target)
     return ret;
 }
 
-void NodeTable::ping(NodeEntry const& _nodeEntry, boost::optional<NodeID> const& _replacementNodeID)
+void NodeTable::ping(Node const& _node, shared_ptr<NodeEntry> _replacementNodeEntry)
+{
+    // Don't sent Ping if one is already sent
+    if (contains(m_sentPings, _node.id))
+    {
+        LOG(m_logger) << "Ignoring request to ping " << _node << ", because it's already pinged";
+        return;
+    }
+
+    PingNode p{m_hostNodeEndpoint, _node.endpoint};
+    p.ts = nextRequestExpirationTime();
+    auto const pingHash = p.sign(m_secret);
+    LOG(m_logger) << p.typeName() << " to " << _node;
+    m_socket->send(p);
+
+    m_sentPings[_node.id] = {
+        _node.endpoint, chrono::steady_clock::now(), pingHash, move(_replacementNodeEntry)};
+}
+
+void NodeTable::schedulePing(Node const& _node)
 {
-    m_timers.schedule(0, [this, _nodeEntry, _replacementNodeID](
-                             boost::system::error_code const& _ec) {
+    m_timers.schedule(0, [this, _node](boost::system::error_code const& _ec) {
         if (_ec || m_timers.isStopped())
             return;
 
-        NodeIPEndpoint src;
-        src = m_hostNodeEndpoint;
-        PingNode p(src, _nodeEntry.endpoint);
-        p.ts = nextRequestExpirationTime();
-        auto const pingHash = p.sign(m_secret);
-        LOG(m_logger) << p.typeName() << " to " << _nodeEntry.id << "@" << p.destination;
-        m_socket->send(p);
-
-        m_sentPings[_nodeEntry.id] = {chrono::steady_clock::now(), pingHash, _replacementNodeID};
-        if (m_nodeEventHandler && _replacementNodeID)
-            m_nodeEventHandler->appendEvent(_nodeEntry.id, NodeEntryScheduledForEviction);
+        ping(_node, {});
     });
 }
 
-void NodeTable::evict(NodeEntry const& _leastSeen, NodeEntry const& _new)
+void NodeTable::evict(NodeEntry const& _leastSeen, shared_ptr<NodeEntry> _replacement)
 {
     if (!m_socket->isOpen())
         return;
 
-    ping(_leastSeen, _new.id);
+    LOG(m_logger) << "Evicting node " << static_cast<Node const&>(_leastSeen);
+    ping(_leastSeen, std::move(_replacement));
+
+    if (m_nodeEventHandler)
+        m_nodeEventHandler->appendEvent(_leastSeen.id, NodeEntryScheduledForEviction);
 }
 
-void NodeTable::noteActiveNode(Public const& _pubk, bi::udp::endpoint const& _endpoint)
+void NodeTable::noteActiveNode(shared_ptr<NodeEntry> _nodeEntry, bi::udp::endpoint const& _endpoint)
```

Comment:
The changes in this method are only:
- `shared_ptr<NodeEntry>` instead of `Public` paramter.
- insert into `m_allNodes` together with inserting into the bucket